test/rateLimiters/tokenBucket.test.ts

```diff
@@ -0,0 +1,265 @@
+import redis from 'redis-mock';
+import { RedisClientType } from 'redis';
+import TokenBucket from '../../src/rateLimiters/tokenBucket';
+
+const CAPACITY = 10;
+// FIXME: Changing the refill rate effects test outcomes.
+const REFILL_RATE = 1; // 1 token per second
+
+let limiter: TokenBucket;
+let client: RedisClientType;
+let timestamp: number;
+const user1 = '1';
+const user2 = '2';
+const user3 = '3';
+const user4 = '4';
+
+async function getBucketFromClient(
+    redisClient: RedisClientType,
+    uuid: string
+): Promise<RedisBucket> {
+    return redisClient.get(uuid).then((res) => JSON.parse(res || '{}'));
+}
+
+async function setTokenCountInClient(
+    redisClient: RedisClientType,
+    uuid: string,
+    tokens: number,
+    time: number
+) {
+    const value: RedisBucket = { tokens, timestamp: time };
+    await redisClient.set(uuid, JSON.stringify(value));
+}
+
+xdescribe('Test TokenBucket Rate Limiter', () => {
+    beforeEach(async () => {
+        // Initialize a new token bucket before each test
+        // create a mock user
+        // intialze the token bucket algorithm
+        client = redis.createClient();
+        limiter = new TokenBucket(CAPACITY, REFILL_RATE, client);
+        timestamp = new Date().valueOf();
+    });
+
+    describe('TokenBucket returns correct number of tokens and updates redis store as expected', () => {
+        test('after an ALLOWED request', async () => {
+            // Bucket intially full
+            const withdraw5 = 5;
+            expect(limiter.processRequest(user1, timestamp, withdraw5).tokens).toBe(
+                CAPACITY - withdraw5
+            );
+            const tokenCountFull = await getBucketFromClient(client, user1);
+            expect(tokenCountFull).toBe(CAPACITY - withdraw5);
+
+            // Bucket partially full but enough time has elapsed to fill the bucket since the last request and
```

Comment:
Break these out into separate tests instead of having one test for all of these variations on situations with allowed requests